src/manifolds/Hamiltonian.jl

```diff
@@ -0,0 +1,210 @@
+[`HamiltonianMatrices`](@ref) `M`, i.e. `X` has to be a Hamiltonian matrix
+The tolerance for [`is_hamiltonian`](@ref) `X` can be set using `kwargs...`.
+"""
+function check_vector(M::HamiltonianMatrices, p, X; kwargs...)
+    if !is_hamiltonian(X; kwargs...)
+        return DomainError(
+            norm((Hamiltonian(X)^+).value + X),
+            "The vector $(X) is not a tangent vector to $(p) on $(M), since it is not hamiltonian.",
+        )
+    end
+    return nothing
+end
+
+embed(::HamiltonianMatrices, p) = p
+embed(::HamiltonianMatrices, p, X) = X
+
+function get_embedding(::HamiltonianMatrices{TypeParameter{Tuple{N}},𝔽}) where {N,𝔽}
+    return Euclidean(2 * N, 2 * N; field=𝔽)
+end
+function get_embedding(M::HamiltonianMatrices{Tuple{Int},𝔽}) where {𝔽}
+    N = get_parameter(M.size)[1]
+    return Euclidean(2 * N, 2 * N; field=𝔽, parameter=:field)
+end
+
+"""
+    is_flat(::HamiltonianMatrices)
+
+Return true. [`HamiltonianMatrices`](@ref) is a flat manifold.
+"""
+is_flat(M::HamiltonianMatrices) = true
+
+@doc raw"""
+    is_hamiltonian(A; kwargs...)
+
+Test whether a matrix `A` is hamiltonian.
+The test consists of verifying whether
+
+```math
+A^+ = -A
+```
+where ``A^+`` denotes the [`symplectic_inverse`](@ref) of `A`.
+
+The passed keyword arguments are passed on to `isapprox`
+check within
+"""
+function is_hamiltonian(A::AbstractMatrix; kwargs...)
+    return isapprox(symplectic_inverse(A), -A; kwargs...)
+end
+function is_hamiltonian(A::Hamiltonian; kwargs...)
+    return isapprox((A^+).value, -A.value; kwargs...)
+end
+
+function Base.show(io::IO, ::HamiltonianMatrices{TypeParameter{Tuple{n}},F}) where {n,F}
+    return print(io, "HamiltonianMatrices($(2n), $(F))")
+end
+function Base.show(io::IO, M::HamiltonianMatrices{Tuple{Int},F}) where {F}
+    n = get_parameter(M.size)[1]
+    return print(io, "HamiltonianMatrices($(2n), $(F); parameter=:field)")
+end
+
+@doc raw"""
+    p = rand(M::HamiltonianMatrices; σ::Real=1.0, vector_at=nothing)
+    rand(M::HamiltonianMatrices; σ::Real=1.0, vector_at=nothing)
+
+Generate a Hamiltonian matrix. Since these are a submanifold of ``ℝ^{2n×2n}``,
+the same method applies for points and tangent vectors.
+
+The generation is based on generating one normally-distributed
+``n×n`` matrix ``A`` and two symmetric ``n×n`` matrices ``B,C`` to generate
+
+```math
+p = \begin{pmatrix} A & B\\ C & -A^{\mathrm{T}} \end{pmatrix}
+```
+
+"""
+rand(M::HamiltonianMatrices; σ::Real=1.0)
+
+function rand!(
+    rng::AbstractRNG,
+    M::HamiltonianMatrices{<:Any,ℝ},
+    pX;
+    σ::Real=one(real(eltype(pX))),
+    vector_at=nothing,
+)
+    n = get_parameter(M.size)[1]
+    p1 = @view(pX[1:n, 1:n])
+    p2 = @view(pX[1:n, (n + 1):(2n)])
+    p3 = @view(pX[(n + 1):(2n), 1:n])
+    p4 = @view(pX[(n + 1):(2n), (n + 1):(2n)])
+    randn!(rng, p1)
+    p4 .= -p1'
+    randn!(rng, p2)
+    randn!(rng, p3)
+    p2 .= (1 / 2) .* (p2 .+ p2')
+    p3 .= (1 / 2) .* (p2 .+ p2')
+    pX .= (σ / norm(pX, 2)) .* pX
```

Comment:
Isn't sigma supposed to dictate standard deviation of the distribution, not the exact norm of the matrix? I'd just multiply ``pX`` by sigma.

Reply:
I think I really just rewrote the existing formula.

Reply:
Then the existing formula was also incorrect.

Reply:
Maybe, I can check and fix that.

Reply:
Fixed it as you proposed.